In [35]:
import pandas as pd 

In [36]:
df=pd.read_csv("iris.csv")
df

,id,SepalLengthCm,SepalWidthCm,PetalLengthCm,PetalWidthCm,Species
0,1,5.1,3.5,1.4,0.2,Iris-setosa
1,2,4.9,3.0,1.4,0.2,Iris-setosa
2,3,4.7,3.2,1.3,0.2,Iris-setosa
3,4,4.6,3.1,1.5,0.2,Iris-setosa
4,5,5.0,3.6,1.4,0.2,Iris-setosa
...,...,...,...,...,...,...
145,146,6.7,3.0,5.2,2.3,Iris-virginica
146,147,6.3,2.5,5.0,1.9,Iris-virginica
147,148,6.5,3.0,5.2,2.0,Iris-virginica
148,149,6.2,3.4,5.4,2.3,Iris-virginica


In [37]:
x=df.drop(["Species"],axis=1)
x

,id,SepalLengthCm,SepalWidthCm,PetalLengthCm,PetalWidthCm
0,1,5.1,3.5,1.4,0.2
1,2,4.9,3.0,1.4,0.2
2,3,4.7,3.2,1.3,0.2
3,4,4.6,3.1,1.5,0.2
4,5,5.0,3.6,1.4,0.2
...,...,...,...,...,...
145,146,6.7,3.0,5.2,2.3
146,147,6.3,2.5,5.0,1.9
147,148,6.5,3.0,5.2,2.0
148,149,6.2,3.4,5.4,2.3


In [38]:
y=df["Species"]
y

0         Iris-setosa
1         Iris-setosa
2         Iris-setosa
3         Iris-setosa
4         Iris-setosa
            ...      
145    Iris-virginica
146    Iris-virginica
147    Iris-virginica
148    Iris-virginica
149    Iris-virginica
Name: Species, Length: 150, dtype: object

In [39]:
from sklearn.model_selection import train_test_split

In [40]:
x_train,x_test,y_train,y_test=train_test_split(x,y,train_size=0.8,random_state=42)

knn

In [41]:
from sklearn.neighbors import KNeighborsClassifier

In [42]:
knn=KNeighborsClassifier()

In [43]:
knn.fit(x_train,y_train)

KNeighborsClassifier()

In [51]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

In [52]:
train_pred=knn.predict(x_train)
train_acc= accuracy_score(y_train,train_pred)
print('accuracy_score',train_acc)

accuracy_score 1.0


In [53]:
test_pred=knn.predict(x_test)
test_acc=accuracy_score(y_test,test_pred)       
print('accuracy_score',test_acc)

accuracy_score 1.0


In [61]:
train_pre = precision_score(y_train, train_pred, average='macro')
print('precision_score', train_pre)

train_rec = recall_score(y_train, train_pred, average='macro')
print('recall_score', train_rec)

train_f1 = f1_score(y_train, train_pred, average='macro')
print('f1_score', train_f1)

precision_score 1.0
recall_score 1.0
f1_score 1.0


In [79]:
test_pre = precision_score(y_test, test_pred, average='macro')
print('precision_score', test_pre)

test_rec = recall_score(y_test, test_pred, average='macro')
print('recall_score', test_rec)

test_f1 = f1_score(y_test, test_pred, average='macro')
print('f1_score', test_f1)

precision_score 1.0
recall_score 1.0
f1_score 1.0



naive_bayes


In [64]:
from sklearn.naive_bayes import GaussianNB

In [65]:
nv=GaussianNB()

In [66]:
nv.fit(x_train,y_train)

GaussianNB()

In [67]:
train_pred=nv.predict(x_train)

In [68]:
test_pred=knn.predict(x_test)

In [70]:
train_acc= accuracy_score(y_train,train_pred)
print('accuracy_score',train_acc)

train_pre = precision_score(y_train, train_pred, average='macro')
print('precision_score', train_pre)

train_rec = recall_score(y_train, train_pred, average='macro')
print('recall_score', train_rec)

train_f1 = f1_score(y_train, train_pred, average='macro')
print('f1_score', train_f1)

accuracy_score 0.9916666666666667
precision_score 0.9920634920634921
recall_score 0.9914529914529915
f1_score 0.9916549314139674


In [72]:
test_acc=accuracy_score(y_test,test_pred)       
print('accuracy_score',test_acc)

test_pre = precision_score(y_test, test_pred, average='macro')
print('precision_score', test_pre)

test_rec = recall_score(y_test, test_pred, average='macro')
print('recall_score', test_rec)

test_f1 = f1_score(y_test, test_pred, average='macro')
print('f1_score', test_f1)

accuracy_score 1.0
precision_score 1.0
recall_score 1.0
f1_score 1.0


DecisionTree

In [73]:
from sklearn.tree import DecisionTreeClassifier

In [74]:
tree=DecisionTreeClassifier()

In [75]:
tree.fit(x_train,y_train)

DecisionTreeClassifier()

In [76]:
train_pred=nv.predict(x_train)
test_pred=knn.predict(x_test)

In [77]:
train_acc= accuracy_score(y_train,train_pred)
print('accuracy_score',train_acc)

train_pre = precision_score(y_train, train_pred, average='macro')
print('precision_score', train_pre)

train_rec = recall_score(y_train, train_pred, average='macro')
print('recall_score', train_rec)

train_f1 = f1_score(y_train, train_pred, average='macro')
print('f1_score', train_f1)

accuracy_score 0.9916666666666667
precision_score 0.9920634920634921
recall_score 0.9914529914529915
f1_score 0.9916549314139674


In [78]:
test_acc=accuracy_score(y_test,test_pred)       
print('accuracy_score',test_acc)

test_pre = precision_score(y_test, test_pred, average='macro')
print('precision_score', test_pre)

test_rec = recall_score(y_test, test_pred, average='macro')
print('recall_score', test_rec)

test_f1 = f1_score(y_test, test_pred, average='macro')
print('f1_score', test_f1)

accuracy_score 1.0
precision_score 1.0
recall_score 1.0
f1_score 1.0


Report

Based on the obtained results, there is a clear difference in performance between the three classifiers.

The KNN model achieved perfect scores (1.0) in all evaluation metrics, including accuracy, precision, recall, and F1-score, in both training and 
testing. This indicates that the model performed consistently well and showed strong stability.

On the other hand, both Naive Bayes and Decision Tree showed slightly lower performance during training, with values around 0.991. However, their
performance improved in the testing phase, where they achieved perfect scores (1.0) across all metrics.

Overall, KNN can be considered the best model in this comparison because it maintained perfect performance in both training and testing, while the 
other models showed minor drops during training.